### PR TITLE
Remove read-only mark on `dist-newstyle` when doing `cabal clean` on Windows

### LINKED ISSUE #10182: Cabal clean fails on Windows when there are s-r-p because of read-only git internal files
**Describe the bug**

These errors are well known on Windows:

```
cabal.test.hs: C:\msys64\tmp\cabal-testsuite-11720\cabal.dist\work\dist\src\bytestring-ea6cd2d51c6ca30b\.git\objects\pack\pack-a66da28598c26cf408a3e8bea5443af982623ae2.rev: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "\\\\?\\C:\\msys64\\tmp\\cabal-testsuite-11720\\cabal.dist\\work\\dist\\src\\bytestring-ea6cd2d51c6ca30b\\.git\\objects\\pack\\pack-a66da28598c26cf408a3e8bea5443af982623ae2.rev": permission denied (Access is denied.)
```
or in particular for `cabal clean`
```
➜ cabal clean
C:\Users\Javier\postCheckoutCommand\dist-newstyle\src\bytestring-ea6cd2d51c6ca30b\.git\objects\pack\pack-8793205445310ec514631949079f7ebb57de3771.rev: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "\\\\?\\C:\\Users\\Javier\\postCheckoutCommand\\dist-newstyle\\src\\bytestring-ea6cd2d51c6ca30b\\.git\\objects\\pack\\pack-8793205445310ec514631949079f7ebb57de3771.rev": permission denied (Access is denied.)
```
and having to do a `rm -rf dist-newstyle` is the only way to clean up the directory.

I have been digging a bit into this and it turns out that those git files are read-only which in Windows need to be forcibly removed and on Linux they can be removed right away:

```
# Linux
javier@EVA-01:~/ghcup-hs/.git/objects/pack$ ls -lah
total 11M
drwxr-xr-x 2 javier javier 4.0K Jul  6 17:41 .
drwxr-xr-x 4 javier javier 4.0K Jul  6 17:40 ..
-r--r--r-- 1 javier javier 370K Jul  6 17:41 pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.idx
-r--r--r-- 1 javier javier 9.9M Jul  6 17:41 pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.pack
-r--r--r-- 1 javier javier  53K Jul  6 17:41 pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.rev
javier@EVA-01:~/ghcup-hs/.git/objects/pack$ ghci
GHCi, version 9.8.2: https://www.haskell.org/ghc/  :? for help
ghci> import System.Directory
ghci> removeFile "pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.rev"
ghci>
Leaving GHCi.
javier@EVA-01:~/ghcup-hs/.git/objects/pack$ ls
pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.idx  pack-f4a5e478e761108dfcb80e76016efbfdcb0ef126.pack
```

```
# Windows
➜ ls -lah aa
-r--r--r-- 1 Javier Javier 0 Jul  6 16:22 aa

~
➜ ghci
GHCi, version 9.8.2: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from C:\Users\Javier\.ghci
λ: import System.Directory
λ: removeFile "aa"
*** Exception: aa: DeleteFile "\\\\?\\C:\\Users\\Javier\\aa": permission denied (Access is denied.)
```

For this reason, whenever there is an s-r-p, a `cabal clean` could fail. Perhaps we should consider making `cabal clean` use `removePathForcibly` instead of `removeDirectoryRecursive`.

### PR DESCRIPTION
**Pull request by jasagredo**
Just one commit, the rest are from the base branch #10255. This still doesn't ensure that the directory can be deleted on the first go, but at least a second cabal clean will for sure clean it.

Closes #10182 #6933

* [X] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [X] The documentation has been updated, if necessary.
* [X] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

Depends-On: #10225 

# QA

Very simple, just doing `cabal clean` if there is a `source-repository-package`
```
❯ cat cabal.project
packages: .

source-repository-package
  type: git
  location: https://github.com/haskell/bytestring

❯ cabal build
Cloning into 'C:\Users\Javier\code\aa\dist-newstyle\src\bytestring-ea6cd2d51c6ca30b'...
...

❯ cabal clean
C:\Users\Javier\code\aa\dist-newstyle\src\bytestring-ea6cd2d51c6ca30b\.git\objects\pack\pack-8793205445310ec514631949079f7ebb57de3771.rev: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "\\\\?\\C:\\Users\\Javier\\code\\aa\\dist-newstyle\\src\\bytestring-ea6cd2d51c6ca30b\\.git\\objects\\pack\\pack-8793205445310ec514631949079f7ebb57de3771.rev": permission denied (Access is denied.)
```

The above will work with this patch.

### CODE REVIEW

**Review by Mikolaj — APPROVED** [review]
Thank you.

**Review by andreabedini — APPROVED** [review]
:+1: I would have never thought deleting files on Windows would be so hard :joy:

**Comment by geekosaur** [comment]
Are we waiting on something or should the "merge me" label be added now?

**Comment by jasagredo** [comment]
I wanted to add a test before merging.

**Comment by jasagredo** [comment]
Added test, we can merge

**Comment by jasagredo** [comment]
Retrying once doesn't seem to solve the issue. Should I:
a) remove the test, we know this thing is flaky but the fix just makes it sometimes less flaky but not always
b) retry until it succeeds
?

**Comment by ulysses4ever** [comment]
This is a driveby opinion. Removing the test would be a pity, so not (a). If there's a slight increase in CI spurious failures with the test enabled (I assume, there is, but just sanity checking), it shouldn't be enabled, so not (b). This leaves me at (c): mark the test as skipped and refer to an issue best positioned to explain the problem.

**Comment by jasagredo** [comment]
When I say the test I mean the new test that I have introduced, which just gets a source-repository-package and then cleans. It is mainly interesting in Windows, so skipping it is kind of not very interesting.

I wonder if another combinator could be created that specifies that a test is flaky, and accepts it whether it passes or not. If it passes many times hopefully someone will eventually notice that it started passing again and maybe re-enabling it can be done at that point.

Or that it specifies the output also for failing and something that says "sometimes it fails, if it fails with this output the it is ok because it is flake, if it fails with another output then there is a different error going on, if it passes then it's fine"

**Comment by jasagredo** [comment]
Depends on #10225

**Comment by geekosaur** [comment]
FWIW you can put that in the top of the pull request: https://docs.mergify.com/workflow/actions/merge/#pull-request-dependencies

**Comment by ulysses4ever** [comment]
> FWIW you can put that in the top of the pull request: [docs.mergify.com/workflow/actions/merge/#pull-request-dependencies](https://docs.mergify.com/workflow/actions/merge/#pull-request-dependencies)

wow, that rocks!

**Comment by geekosaur** [comment]
Is this still waiting on something?

**Comment by tomsmeding** [comment]
For some reason, perhaps because this package has a custom Setup.hs, `cabal build` seems to want to reinitialise submodules of `source-repository-packages` repos very often. Normally this is annoying but not really a problem, but on Windows, due to the issue that this PR addresses, this means that **the second `cabal build` will always fail** for this package. If you have built once, you cannot build again -- you have to delete `dist-newstyle` first, then you can build from scratch, thank you.

Please merge this PR and release it as soon as possible. :)

**Comment by geekosaur** [comment]
@jasagredo, GitHub is complaining about conflicts and Mergify is throwing errors trying to correct them?

**Comment by ulysses4ever** [comment]
Maybe someone other than the author should manually rebase, if the author is currently busy.

**Comment by geekosaur** [comment]
I considered doimng it myself, but since it's Windows I worry that I'd run into things I wouldn't know how to resolve.

**Comment by jasagredo** [comment]
I will rebase and merge this PR.

@tomsmeding keep in mind what I mentioned in the description: This is not a silver bullet. Perhaps what was here in the first version was better, namely `removeDirectoryForcibly`, but that still might fail due to the files being in use 🙄

**Comment by tomsmeding** [comment]
@jasagredo No guarantees I'll have time (sorry, have to find a windows machine again), but to increase the chances: is there a cabal windows binary that I can easily download (CI artifacts?) that includes this PR?

**Comment by geekosaur** [comment]
https://github.com/haskell/cabal/releases/download/cabal-head/cabal-head-Windows-x86_64.tar.gz

**Comment by tomsmeding** [comment]
This works beautifully for me! Thanks for fixing (at least for my usecase); looking forward to the release.

**Comment by Bodigrim** [comment]
> I implemented a slightly stronger version (remove read-only mark + force delete the directory). 

Do we actually need to remove read-only mark if we use `removePathForcibly` anyway?..

**Pull request merged**
